homeassistant/components/screenlogic/data.py

```diff
@@ -0,0 +1,348 @@
+    EntityParameter,
+    Callable | EntityCategory | int | ScreenLogicRule | tuple,
+]
+
+SupportedValueDescriptions = dict[str, SupportedValueParameters | dict]
+
+SupportedGroupDescriptions = dict[int | str, SupportedValueDescriptions]
+
+SupportedDeviceDescriptions = dict[str, SupportedGroupDescriptions]
+
+
+@dataclass
+class BaseScreenLogicEntityData:
+    """Generic representation of a ScreenLogic entity."""
+
+    data_path: tuple[str | int, ...]
+    enabled: bool
+    entity_key: str
+    subscription_code: int | None
+    value_data: dict
+    value_parameters: dict
+
+
+DEVICE_INCLUSION_RULES = {
+    DEVICE.PUMP: ScreenLogicDataRule(
+        lambda pump_data: pump_data[VALUE.DATA] != 0,
+        (PathPart.DEVICE, PathPart.INDEX),
+    ),
+    DEVICE.INTELLICHEM: ScreenLogicEquipmentRule(
+        lambda flags: EQUIPMENT_FLAG.INTELLICHEM in flags,
+    ),
+    DEVICE.SCG: ScreenLogicEquipmentRule(
+        lambda flags: EQUIPMENT_FLAG.CHLORINATOR in flags,
+    ),
+}
+
+DEVICE_SUBSCRIPTION = {
+    DEVICE.CONTROLLER: CODE.STATUS_CHANGED,
+    DEVICE.INTELLICHEM: CODE.CHEMISTRY_CHANGED,
+}
+
+
+def get_ha_unit(entity_data: dict) -> StrEnum | str | None:
+    """Return a Home Assistant unit of measurement from a UNIT."""
+    sl_unit = entity_data.get(ATTR.UNIT)
+    return SL_UNIT_TO_HA_UNIT.get(sl_unit, sl_unit)
+
+
+def realize_path_template(
+    template_path: tuple[str | int, ...], data_path: tuple[str | int, ...]
+) -> tuple[str | int, ...]:
+    """Make data path from template and current."""
+    if not data_path or len(data_path) < 3:
+        raise KeyError(
+            f"Missing or invalid required parameter: 'data_path' for template path '{template_path}'"
+        )
+    device, group, data_key = data_path
+    realized_path = []
+    for part in template_path:
+        match part:
+            case PathPart.DEVICE:
+                realized_path.append(device)
+            case PathPart.INDEX | PathPart.KEY:
+                realized_path.append(group)
+            case PathPart.VALUE:
+                realized_path.append(data_key)
+            case _:
+                realized_path.append(part)
+
+    return tuple(realized_path)
+
+
+def process_supported_values(
+    coordinator: ScreenlogicDataUpdateCoordinator,
+    platform_domain: str,
+    supported_devices: SupportedDeviceDescriptions,
+):
+    """Process template data."""
+    entity_registry = er.async_get(coordinator.hass)
+
+    def cleanup_excluded_entity(entity_key: str):
+        """Remove entity if it exists."""
+        assert coordinator.config_entry
+        unique_id = f"{coordinator.config_entry.unique_id}_{entity_key}"
+        if entity_id := entity_registry.async_get_entity_id(
+            platform_domain, SL_DOMAIN, unique_id
+        ):
+            _LOGGER.debug(
+                "Removing existing entity '%s' per data inclusion rule", entity_id
+            )
+            entity_registry.async_remove(entity_id)
```

Comment:
This looks like entity logic

Reply:
This was specifically for the Saturation Index issue. As I was already processing if a supported data point should be included or not based on the inclusion rules, this seemed like the logical (and earliest) place to remove any entities that already existed but shouldn't be included anymore.